src/rgw/rgw_torrent.cc

```diff
@@ -0,0 +1,291 @@
+{
+  info.sha1_bl.append(buff, CEPH_CRYPTO_SHA1_DIGESTSIZE);
+}
+
+void seed::set_info_name(string value)
+{
+  info.name = value;
+}
+
+void seed::sha1(SHA1 *h, bufferlist &bl, off_t bl_len)
+{
+  off_t num = bl_len/info.piece_length;
+  off_t remain = 0;
+  remain = bl_len%info.piece_length;
+
+  char *pstr = bl.c_str();
+  char sha[25];
+
+  /* get sha1 */
+  for (off_t i = 0; i < num; i++)
+  {
+    memset(sha, 0x00, sizeof(sha));
+    h->Update((byte *)pstr, info.piece_length);
+    h->Final((byte *)sha);
+    set_info_pieces(sha);
+    pstr += info.piece_length;
+  }
+
+  /* process remain */
+  if (0 != remain)
+  {
+    memset(sha, 0x00, sizeof(sha));
+    h->Update((byte *)pstr, remain);
+    h->Final((byte *)sha);
+    set_info_pieces(sha);
+  }
+}
+
+int seed::sha1_process()
+{
+  uint64_t remain = info.len%info.piece_length;
+  uint8_t  remain_len = ((remain > 0)? 1 : 0);
+  sha_len = (info.len/info.piece_length + remain_len)*CEPH_CRYPTO_SHA1_DIGESTSIZE;
+
+  SHA1 h;
+  list<bufferlist>::iterator iter = torrent_bl.begin();
+  for (; iter != torrent_bl.end(); iter++)
+  {
+    bufferlist &bl_info = *iter;
+    sha1(&h, bl_info, (*iter).length());
+  }
+
+  return 0;
+}
+
+int seed::handle_data()
+{
+  int ret = 0;
+
+  /* sha1 process */
+  ret = sha1_process();
+  if (0 != ret)
+  {
+    ldout(s->cct, 0) << "ERROR: failed to sha1_process() ret= "<< ret << dendl;
+    return ret;
+  }
+
+  /* produce torrent data */
+  do_encode();
+
+  /* save torrent data into OMAP */
+  ret = save_torrent_file();
+  if (0 != ret)
+  {
+    ldout(s->cct, 0) << "ERROR: failed to save_torrent_file() ret= "<< ret << dendl;
+    return ret;
+  }
+
+  return 0;
+}
+
+int seed::get_params()
+{
+  is_torrent = true;
+  info.piece_length = g_conf->rgw_torrent_sha_unit;
+  create_by = g_conf->rgw_torrent_createby;
+  encoding = g_conf->rgw_torrent_encoding;
+  origin = g_conf->rgw_torrent_origin;
+  comment = g_conf->rgw_torrent_comment;
+  announce = g_conf->rgw_torrent_tracker;
+
+  /* tracker and tracker list is empty, set announce to origin */
+  if (announce.empty() && !origin.empty())
+  {
+    announce = origin;
+  }
+
+  return 0;
+}
+
+void seed::set_announce()
+{
+  const char *pstr = announce.c_str();
+  char *pstart = const_cast<char *>(pstr);
+  uint64_t len = announce.length();
+  char *pend = pstart + len;
+  char *location;
+  char flag = ',';
+
+  list<string> announce_list;  // used to get announce list from conf
+  while (pstart < pend)
+  {
+    location = NULL;
+    location = strchr(pstart, flag);
+    if (!location)
+    {
+      string ann(pstart);
+      announce_list.push_back(ann);
+      break;
+    }
+
+    char temp[100] = { 0 };
+    snprintf(temp, location - pstart + 1, "%s", pstart);
+    string anno(temp);
+    announce_list.push_back(anno);
+    pstart = location + 1;
+  }
+
+  list<string>::iterator iter = announce_list.begin();
+  dencode.bencode_key(ANNOUNCE, bl);
+  dencode.bencode_key((*iter), bl);
+  iter++;
```

Comment:
this `iter++` doesn't look right. ANNOUNCE_LIST should contain all of the urls, but you're skipping the first here